runtime/jvm/src/main/kotlin/kotlinx/serialization/Serialization.kt

```diff
@@ -35,12 +35,24 @@ actual fun <T: Any, E: T?> ArrayList<E>.toNativeArray(eClass: KClass<T>): Array<
 
 @Suppress("UNCHECKED_CAST")
 internal fun <T> Class<T>.invokeSerializerGetter(vararg args: KSerializer<Any>): KSerializer<T>? {
-    val companion: Any = try {
-        this.getField("Companion").apply { isAccessible = true }.get(null)
-    } catch (e: NoSuchFieldException) {
-        return null
-    }
-    return companion.javaClass.methods
-            .find {it.name == "serializer" && it.parameterTypes.size == args.size && it.parameterTypes.all {it == KSerializer::class.java } }
+    var serializer: KSerializer<T>? = null
+
+    // Search for serializer defined on companion object.
+    val companion = declaredFields.singleOrNull {it.name == "Companion"}?.apply {isAccessible = true}?.get(null)
```

Comment:
I wonder if iterating all fields will not slow down lookup dramatically. @elizarov what do you think?

Reply:
The alternative would be catching the exception thrown by `getField` to handle this logic, which is very likely to be slower.

Most likely under the covers `getField()` traverses the fields just as well. Doubt this would impact performance very much. I chose `singleOrNull` over `firstOrNull` since it seemed 'unexpected behavior to me in case two companion objects would be defined. But unfortunately I am not familiar with the specification, so this might be a totally wrong presumption.

Reply:
Apparently you can have only one companion object. I get a warning in the editor when trying to add a second. Oddly enough, [this does not seem documented](https://kotlinlang.org/docs/reference/object-declarations.html), so perhaps my idea there could be several was not too far-fetched.

In case performance _would_ be an issue, the result of this code using reflection (which by itself typically incurs an impact) should likely be cached. Functionality first, optimization second. :)